main.go

```diff
@@ -224,161 +119,355 @@ func addEnvVars(ciProvider string) {
+		// https://github.com/spf13/viper/issues/461#issuecomment-366831834
+		root.PersistentFlags().Lookup("apikey").Value.Set(apikey)
 	}
-	if apihost == "" {
-		apihost = "https://api.honeycomb.io"
+
+	root.PersistentFlags().StringVarP(&config.Dataset, "dataset", "d", "buildevents", "[env.BUILDEVENT_DATASET] the name of the Honeycomb dataset to which to send these events")
+	if dataset, ok := os.LookupEnv("BUILDEVENT_DATASET"); ok {
+		root.PersistentFlags().Lookup("dataset").Value.Set(dataset)
 	}
-	// use timeout default of 10min
-	if timeoutStr == "" {
-		timeoutStr = "10"
+
+	root.PersistentFlags().StringVarP(&config.APIHost, "apihost", "a", "https://api.honeycomb.io", "[env.BUILDEVENT_APIHOST] the hostname for the Honeycomb API server to which to send this event")
+	if apihost, ok := os.LookupEnv("BUILDEVENT_APIHOST"); ok {
+		root.PersistentFlags().Lookup("apihost").Value.Set(apihost)
 	}
-	timeoutMin, err := strconv.Atoi(timeoutStr)
-	if err != nil {
-		timeoutMin = 10
+
+	root.PersistentFlags().StringVarP(&filename, "filename", "f", "", "[env.BUILDEVENT_FILE] the path of a text file holding arbitrary key=val pairs (multi-line-capable, logfmt style) to be added to the Honeycomb event")
+	if fname, ok := os.LookupEnv("BUILDEVENT_FILE"); ok {
+		root.PersistentFlags().Lookup("filename").Value.Set(fname)
 	}
 
-	if Version == "" {
-		Version = "dev"
+	root.PersistentFlags().StringVarP(&ciProvider, "provider", "p", "", "[env.BUILDEVENT_CIPROVIDER] if unset, will inspect the environment to try and detect Travis-CI, CircleCI, or GitLab-CI.")
+	prov := os.Getenv("BUILDEVENT_CIPROVIDER")
+	if prov == "" {
+		if _, present := os.LookupEnv("TRAVIS"); present {
+			prov = providerTravis
+		} else if _, present := os.LookupEnv("CIRCLECI"); present {
+			prov = providerCircle
+		} else if _, present := os.LookupEnv("GITLAB_CI"); present {
+			prov = providerGitLab
+		}
+	}
+	if prov != "" {
+		root.PersistentFlags().Lookup("provider").Value.Set(prov)
 	}
 
-	// respond to ./buildevents --version
-	if strings.TrimSpace(os.Args[1]) == "--version" {
-		fmt.Println(Version)
-		os.Exit(0)
+	// BUILD - eg: buildevents build $TRAVIS_BUILD_ID $BUILD_START success
+	buildCmd := &cobra.Command{
+		Use:   "build [flags] BUILD_ID BUILD_START OUTCOME",
+		Short: "Sends the root span for the entire build",
+		Long: `
+The build mode sends the root span for the entire build. It should be called
+when the build finishes and records the duration of the entire build. It emits
+a URL pointing to the generated trace in Honeycomb to STDOUT.`,
+		Args:                  argOptions(2, "success", "failure"),
+		DisableFlagsInUseLine: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			ev := createEvent(config, ciProvider, args[0])
+			defer ev.Send()
+
+			providerInfo(ciProvider, ev)
+			arbitraryFields(filename, ev)
+
+			ts, dur := parseSince(args[1])
+			ev.Add(map[string]interface{}{
+				"service_name":  "build",
+				"trace.span_id": args[0],
```

Comment:
I would really like to get rid of `arg[#]` parsing buried in these spots. I can understand not wanting to make the change away from positional arguments right now. Could you move these to the head of the function at least, so it's clear what is coming in? I also see something in the Cobra docs about positional arguments. Perhaps we can leverage that to make this a little cleaner.

Reply:
Cleaning this up!